### ISSUE TITLE
Check On Save doesn't work with clippy?

### ISSUE THREAD

**Issue by DianaNites** [rust-analyzer/rust-analyzer#3871]
I've got the command set to clippy, and the [status bar](https://github.com/rust-analyzer/rust-analyzer/blob/3c9e9d3f3ebbc7a22d932dd2a3fd63f1e44c4568/editors/code/src/status_display.ts#L10) [shows](https://user-images.githubusercontent.com/5275194/78586156-06a90380-7809-11ea-9377-85c7a72c7d2e.png) that clippy is being run. I tried looking at the logs/traces, but the exact commands being run don't seem to be logged.

```json
"rust-analyzer.checkOnSave.command": "clippy"
```

Yet the following code produces no errors from rust-analyzer, even though it should trigger (a false positive) [clippy::cast_ptr_alignment](https://rust-lang.github.io/rust-clippy/master/index.html#cast_ptr_alignment)

This *only* happens if rust-analyzer ran the command. Running `cargo clean && cargo clippy` will properly display the error, saving the file and triggering rust-analyzer running it, will not, and subsequent `cargo clippy` runs will not either.

It was suggested on twitter this may be an interaction between `cargo check` and `cargo clippy`, since the latter will cache the former's lack of errors. This despite the fact rust-analyzer is set to run, and displays as running, `clippy`?

```rust
#[derive(Copy, Clone)]
#[repr(C)]
struct Struct {
    _dummy: u8,
    _dummy2: u32,
}

fn main() {
    let bytes = &[0u8; 100];
    let s = unsafe { (bytes.as_ptr() as *const Struct).read_unaligned() };
}
```

----

Using VSCode 1.43.2
Rust Analyzer commit c859a6480af7baece2eec38c19f71cb714db9e3b

**Comment by JoshMcguigan** [comment]
I'm seeing the same thing. I'm not sure why yet, but I've found setting `"rust-analyzer.checkOnSave.overrideCommand": ["cargo", "clippy", "--workspace", "--message-format=json", "--all-targets"]` does somehow work around the issue.
